lib/Transforms/ShapeInference/ShapeInference.cpp

```diff
@@ -0,0 +1,187 @@
+
+    // Update types with annotations (where it exists)
+    bool changed = false;
+    for (auto &[idx, type] : dynTensorArgs) {
+      // Check if the current argument has any attributes.
+      auto dictattr = dyn_cast<DictionaryAttr>(attrs[idx]);
+      if (!dictattr) continue;  // nothing to infer from
+
+      // Get the shape attribute if it exists
+      auto *shapeAttr = llvm::find_if(dictattr, [&](NamedAttribute attr) {
+        return attr.getName() == "shape.shape";
+      });
+      if (!shapeAttr) continue;  // nothing to infer from
+
+      // Extract the actual shape from the shape annotation attribute
+      auto arrayAttr = dyn_cast<ArrayAttr>(shapeAttr->getValue());
+      if (!arrayAttr)
+        return emitError(funcOp.getLoc(),
+                         "Invalid shape attribute (expected ArrayAttr)");
+
+      if (llvm::any_of(arrayAttr,
+                       [](Attribute attr) { return !isa<IntegerAttr>(attr); }))
+        return emitError(
+            funcOp.getLoc(),
+            "Invalid shape attribute (expected ArrayAttr of only IntegerAttr)");
+
+      SmallVector<int64_t> shape = llvm::to_vector(
+          llvm::map_range(arrayAttr.getAsRange<IntegerAttr>(),
+                          [](IntegerAttr i) { return i.getInt(); }));
+
+      if (shape.size() != type.getRank())
+        return emitError(
+            funcOp.getLoc(),
+            llvm::formatv("Shape annotation has wrong number of elements "
+                          "(expected {0}, got {1}).",
+                          type.getRank(), shape.size()));
+
+      // save the updated argument type
+      type = RankedTensorType::get(shape, type.getElementType());
+      changed = true;
+    }
+    // no new information found to infer from -> exit
+    if (!changed) return failure();
+
+    rewriter.startOpModification(funcOp);
+
+    /////////////////////////////////////////////////////////////////////
+    ////  Step 2: Update Function Type/Signature based on Attributes ////
+    /////////////////////////////////////////////////////////////////////
+    // We need to update the function type to reflect the new types,
+    // update the type of the "real" arguments (block arguments),
+    // and we also remove the attribute argument from the signature
+    // (primarily for better readability of the IR after shape inference).
+    //
+    // We could delay updating the function type until the end,
+    // when we also know the inferred return type(s),
+    // but that would leave the func op in a less consistent state longer
+    auto newInputTypes = llvm::to_vector(funcOp.getFunctionType().getInputs());
+    for (auto [idx, type] : dynTensorArgs) {
+      funcOp.getArgument(idx).setType(type);
+      funcOp.removeArgAttr(idx, "shape.shape");
+      newInputTypes[idx] = type;
+    }
+    auto newFunctionType = rewriter.getFunctionType(
+        newInputTypes, funcOp.getFunctionType().getResults());
+    funcOp.setFunctionType(newFunctionType);
```

Comment:
Any reason to do this here instead of merging it with the equivalent call on line 165-166? Since the block argument types are already changed it shouldn't affect the type inference step 3.